model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java

```diff
@@ -697,6 +696,7 @@ public SessionEntityWrapper<UserSessionEntity> importUserSession(UserSessionMode
 
         // Import client sessions
         clientSessionTx.importSessionsConcurrently(realm, clientSessionsById, offline);
+        clientSessionTx.setUserSessionId(clientSessionsById.keySet(), sessionId, offline);
```

Comment:
It feels a bit non-obvious to do this setting of the user session after the import. The results is correct though. 

Thinking a bit more about `AuthenticatedClientSessionEntity#userSessionId`, it should probably be volatile as it is updated by potentially multiple threads in parallel (if it is stored on the local node), and no serialization is used. 

I realize that I added the `userSessionId` back in the time, I probably should have change it back then already. 

WDYT?

Reply:
Yeah, I'll change it to volatile.